gridstatus/nyiso.py

```diff
@@ -64,8 +73,73 @@ class NYISO(ISOBase):
     status_homepage = "https://www.nyiso.com/system-conditions"
     interconnection_homepage = "https://www.nyiso.com/interconnections"
 
+    def _handle_time(
+        self,
+        df: pd.DataFrame,
+        dataset_name: str,
+        groupby: str | None = None,
+    ) -> pd.DataFrame:
+        time_type, interval_duration_minutes = DATASET_INTERVAL_MAP[dataset_name]
+
+        if "Time Stamp" in df.columns:
+            time_stamp_col = "Time Stamp"
+        elif "Timestamp" in df.columns:
+            time_stamp_col = "Timestamp"
+
+        def time_to_datetime(s: pd.Series, dst: str = "infer") -> pd.Series:
+            return pd.to_datetime(s).dt.tz_localize(
+                self.default_timezone,
+                ambiguous=dst,
+            )
+
+        if "Time Zone" in df.columns:
+            dst = df["Time Zone"] == "EDT"
+            df[time_stamp_col] = time_to_datetime(
+                df[time_stamp_col],
+                dst,
+            )
+
+        elif "Name" in df.columns or groupby:
+            groupby = groupby or "Name"
+            # once we group by name, the time series for each group is no longer ambiguous
+            df[time_stamp_col] = df.groupby(groupby, group_keys=False)[
+                time_stamp_col
+            ].apply(
+                time_to_datetime,
+                "infer",
+            )
+        else:
+            df[time_stamp_col] = time_to_datetime(
+                df[time_stamp_col],
+                "infer",
+            )
+
+        df = df.rename(columns={time_stamp_col: "Time"})
+
+        if time_type != "instantaneous":
+            interval_duration = pd.Timedelta(minutes=interval_duration_minutes)
+            if time_type == "start":
+                df["Interval Start"] = df["Time"]
+                df["Interval End"] = df["Interval Start"] + interval_duration
+            elif time_type == "end":
+                df["Interval Start"] = df["Time"] - interval_duration
+                df["Interval End"] = df["Time"]
+                df["Time"] = df["Interval Start"]
+
+            utils.move_cols_to_front(
+                df,
+                ["Time", "Interval Start", "Interval End"],
+            )
+
+        return df
+
     @support_date_range(frequency="MONTH_START")
-    def get_status(self, date, end=None, verbose=False):
+    def get_status(
+        self,
+        date: str | pd.Timestamp | tuple[pd.Timestamp, pd.Timestamp],
+        end: str | pd.Timestamp | tuple[pd.Timestamp, pd.Timestamp] | None = None,
```

Comment:
Since `date` and `end` are arguments to nearly all functions and they all have the same allowed types, how do you feel about defining constants for these types instead of having to write them out for every function?

```python
DATE_TYPE_ANNOTATION = str | pd.Timestamp | tuple[pd.Timestamp, pd.Timestamp]
END_TYPE_ANNOTATION = str | pd.Timestamp | tuple[pd.Timestamp, pd.Timestamp] | None
```

could go in `gridstatus/utils.py`.

Reply:
Yeah, I agree there's a more thorough refactoring of date handling we can do here. That can wait for future work.